File: 1501~2000/1937. Maximum Number of Points with Cost/main.cpp
/** 
 * Source: https://leetcode.com/problems/maximum-number-of-points-with-cost/
 * Date: 2021/7/19
 * Skill: DP
 * Runtime: 
 * Memory Usage: 
 * Time complexity:
 * Space complexity: 
 * Constraints: 
 *      1 <= m, n <= 10^5
 *      0 <= points[r][c] <= 10^5
 */

#include<iostream>
#include<vector>
using namespace std;

typedef struct point{
    int num;
    int offset;
    point(int a, int b):num(a), offset(b){};
}Point;

class Solution {
public:
    long long maxPoints(vector<vector<int>>& points) {
        long long ans = 0;
        vector<vector<Point>> dp(points.size(), vector<Point>(points[0].size(), Point(-1, 0)));
        for (int i = 0; i < points[0].size(); i++) {
            dp[0][i].num = points[0][i];
            dp[0][i].offset = i;
        }
        for (int j = 1; j < points.size(); j++) {
            dp[j][0].num = points[j][0] + dp[j-1][0].num;
            dp[j][0].offset = 0;
        }
        for (int i = 1; i < points.size(); i++) {
            for (int j = 1; j < points[0].size(); j++) {
                if (dp[i-1][j].num + )
            }
        }
    }
};

static const auto io_sync_off = []() {
    //* turn off sync
    std::ios::sync_with_stdio(false);
    //* untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 1501~2000/1937. Maximum Number of Points with Cost/main.py
"""
Source: https://leetcode.com/problems/maximum-number-of-points-with-cost/
Date: 2023/4/15
Skill:
Ref:
Runtime: 2431 ms, faster than 21.49%
Memory Usage: 47.7 MB, less than 70.9%
Time complexity:
Space complexity:
Constraints:

"""

import math
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        row, col = len(points), len(points[0])
        dp = [[0 for _ in range(col)] for _ in range(row)]
        for i in range(col):
            dp[0][i] = points[0][i]

        for i in range(1, row):
            rolling_max = -math.inf
            for j in range(col):
                rolling_max = max(rolling_max, dp[i - 1][j] + j)
                dp[i][j] = points[i][j] - j + rolling_max

            rolling_max = -math.inf
            for j in range(col - 1, -1, -1):
                rolling_max = max(rolling_max, dp[i - 1][j] - j)
                dp[i][j] = max(dp[i][j], points[i][j] + j + rolling_max)

        return max(dp[row - 1])


if __name__ == "__main__":
    s = Solution()
    points = [[1, 2, 3], [1, 5, 1], [3, 1, 1]]
    res = s.maxPoints(points)
